packages/hash/frontend/src/blocks/userBlocks.tsx

```diff
@@ -11,64 +8,29 @@ import {
   useMemo,
   useState,
 } from "react";
+import { fetchBlockMeta } from "@hashintel/hash-shared/blockMeta";
 
-import { isProduction } from "../lib/config";
 import { useCachedDefaultState } from "../components/hooks/useDefaultState";
-
-export type RemoteBlockMetadata = BlockMetadata & {
-  componentId: string;
-};
-
-type UserBlocks = RemoteBlockMetadata[];
+import { BlocksMetaMap } from "./page/createEditorView";
 
 interface UserBlocksContextState {
-  value: UserBlocks;
-  setValue: Dispatch<SetStateAction<UserBlocks>>;
+  value: BlocksMetaMap;
+  setValue: Dispatch<SetStateAction<BlocksMetaMap>>;
   blockFetchFailed: boolean;
 }
 
 /** @private enforces use of custom provider */
 const UserBlocksContext = createContext<UserBlocksContextState | null>(null);
 
-const mergeBlocksData = (
-  oldBlocksData: UserBlocks,
-  newBlocksData: UserBlocks,
-): UserBlocks => {
-  return produce(oldBlocksData, (draftUserBlocks) => {
-    for (const latestUserBlock of newBlocksData) {
-      const matchingUserBlockIndex = draftUserBlocks.findIndex(
-        (userBlock) => userBlock.name === latestUserBlock.name,
-      );
-
-      // @todo Remove need for @ts-expect-error here
-      if (matchingUserBlockIndex === -1) {
-        // @ts-expect-error TS warns `Type instantiation is excessively deep` but this isn't a problem here
-        draftUserBlocks.push(latestUserBlock);
-      } else if (
-        // in development, overwrite the locally cached block if the source has changed
-        (!isProduction &&
-          draftUserBlocks[matchingUserBlockIndex]?.source !==
-            latestUserBlock.source) ||
-        // overwrite the locally cached block if loading a different version
-        draftUserBlocks[matchingUserBlockIndex]?.version !==
-          latestUserBlock.version
-      ) {
-        // @ts-expect-error TS warns `Type instantiation is excessively deep` but this isn't a problem here
-        draftUserBlocks[matchingUserBlockIndex] = latestUserBlock;
-      }
-    }
-  });
-};
-
-export const UserBlocksProvider: FC<{ value: UserBlocks }> = ({
+export const UserBlocksProvider: FC<{ value: BlocksMetaMap }> = ({
   value: initialUserBlocks,
   children,
 }) => {
   const [value, setValue] = useCachedDefaultState(
     initialUserBlocks,
     "hash-workspace-user-blocks",
     (nextInitialItems, prevInitialItems) => {
-      return mergeBlocksData(prevInitialItems, nextInitialItems);
+      return { ...prevInitialItems, ...nextInitialItems };
```

Comment:
I don't think we need the previous complicated logic, it's fine to override existing blocks with new versions in all cases